[aot] Keep backward mutations in backward

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129130

https://github.com/pytorch/pytorch/issues/127561

Mutations of inputs in backward are emitted manually, after joint_fn tracing.
With default partitioner logic they will be moved to "forward" graph, as this is operation on forward inputs.

To keep those mutations in backward:
- Introduce "subgraph" node key, that can be specified with contextmanager. When we do manual `copy_` in backward on forward input - we know that his is for backward - set subgraph="backward"


In partitioner:
Introducing optional argument subgraph, to filter out nodes with specified subgraph (node_subgraph) and not to add them to subgraph if node_subgraph is different.

IvanKobzarev committed:
[aot] Keep backward mutations in backward

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aot] Keep backward mutations in backward"


https://github.com/pytorch/pytorch/issues/127561

Mutations of inputs in backward are emitted manually, after joint_fn tracing.
With default partitioner logic they will be moved to "forward" graph, as this is operation on forward inputs.

To keep those mutations in backward:
- Introduce "subgraph" node key, that can be specified with contextmanager. When we do manual `copy_` in backward on forward input - we know that his is for backward - set subgraph="backward"


In partitioner:
Introducing optional argument subgraph, to filter out nodes with specified subgraph (node_subgraph) and not to add them to subgraph if node_subgraph is different.


[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -329,6 +329,20 @@ def traced_forward(*primals_fwd_seed_fwd_base_offset):
+@contextmanager
+def set_subgraph(subgraph: str):
+    meta_key = "subgraph"
+    if fx_traceback.has_preserved_node_meta():
```
This is saying that if `has_preserved_node_meta() == False`, we will just silently not add any metadata here (and fail to tell the partitioner that our mutation was a backward mutation, giving us wrong results).

Since we're relying on metadata for this correctness, we should probably just assert that this is true?

bdhirsh commented on torch/fx/proxy.py:
```diff
@@ -94,6 +94,7 @@ def __exit__(self, *args):
     "recompute",
     "from_node",
     "quantization_tag",
+    "subgraph"
```
"subgraph" is a pretty generic-sounding name. What about something more specific, like "is_backward_mutation"?

IvanKobzarev replied:
I think we can reuse this for other partitioner related actions. I will rename to "partitioner_tag": "mutation_in_backward"

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -144,7 +144,10 @@ def __repr__(self):
+    joint_graph: fx.Graph,
+    inputs: List[fx.Node],
+    outputs: List[fx.Node],
+    subgraph: Optional[str] = None,
```
cc @Chillee for the partitioner changes (the partitioner now needs to know if a given `copy_` node should be partitioned into the fw or bw graph, if it is a "backward mutation")

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -2148,24 +2148,29 @@ def setup_context(ctx, inputs, output):
-        y = compiled_f(x, x1)
-        loss = y.sum()
-        loss.backward()
+        y = compiled_f(x, x1, x2)
```
it looks like this test mutates `x2` in the forward, and `x1` in the backward.

I think that if you really want to confirm that our mutations are showing up in the right places, you should assert the values of `x`, `x1` and `x2` against their references before **and** after running the backward
```
y = compiled_f(x, x1, x2)

# assert that only x2 was mutated during the fw
self.assertEqual(x, x_ref)
self.assertEqual(x1, x1_ref)
self.assertEqual(x2, x2_ref)
self.assertEqual(y, y_ref)

# now run the backward and assert that x1 is mutated properly
y.sum().backward()

self.assertEqual(x, x_ref)
self.assertEqual(x1, x1_ref)
self.assertEqual(x2, x2_ref)
self.assertEqual(y, y_ref)

# and the gradients are correct
self.assertEqual(x.grad, x_ref.grad)
self.assertEqual(x1.grad, x1_ref.grad)
self.assertEqual(x2.grad, x2_ref.grad)
```

IvanKobzarev committed:
Update on "[aot] Keep backward mutations in backward"


https://github.com/pytorch/pytorch/issues/127561

Mutations of inputs in backward are emitted manually, after joint_fn tracing.
With default partitioner logic they will be moved to "forward" graph, as this is operation on forward inputs.

To keep those mutations in backward:
- Introduce "subgraph" node key, that can be specified with contextmanager. When we do manual `copy_` in backward on forward input - we know that his is for backward - set subgraph="backward"


In partitioner:
Introducing optional argument subgraph, to filter out nodes with specified subgraph (node_subgraph) and not to add them to subgraph if node_subgraph is different.


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aot] Keep backward mutations in backward"


https://github.com/pytorch/pytorch/issues/127561

Mutations of inputs in backward are emitted manually, after joint_fn tracing.
With default partitioner logic they will be moved to "forward" graph, as this is operation on forward inputs.

To keep those mutations in backward:
- Introduce "subgraph" node key, that can be specified with contextmanager. When we do manual `copy_` in backward on forward input - we know that his is for backward - set subgraph="backward"


In partitioner:
Introducing optional argument subgraph, to filter out nodes with specified subgraph (node_subgraph) and not to add them to subgraph if node_subgraph is different.


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aot] Keep backward mutations in backward"


https://github.com/pytorch/pytorch/issues/127561

Mutations of inputs in backward are emitted manually, after joint_fn tracing.
With default partitioner logic they will be moved to "forward" graph, as this is operation on forward inputs.

To keep those mutations in backward:
- Introduce "subgraph" node key, that can be specified with contextmanager. When we do manual `copy_` in backward on forward input - we know that his is for backward - set subgraph="backward"


In partitioner:
Introducing optional argument subgraph, to filter out nodes with specified subgraph (node_subgraph) and not to add them to subgraph if node_subgraph is different.


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aot] Keep backward mutations in backward"


https://github.com/pytorch/pytorch/issues/127561

Mutations of inputs in backward are emitted manually, after joint_fn tracing.
With default partitioner logic they will be moved to "forward" graph, as this is operation on forward inputs.

To keep those mutations in backward:
- Introduce "subgraph" node key, that can be specified with contextmanager. When we do manual `copy_` in backward on forward input - we know that his is for backward - set subgraph="backward"


In partitioner:
Introducing optional argument subgraph, to filter out nodes with specified subgraph (node_subgraph) and not to add them to subgraph if node_subgraph is different.


[ghstack-poisoned]

ezyang commented:
@Chillee could you please take a look at this

bdhirsh commented:
Looks good to me minus the partitioner bits (I'll wait to see if Horace can review)

Chillee commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -443,7 +456,11 @@ def _functionalized_f_helper(*args):
                     # we'll check at runtime and fail only when backward is under torch.is_grad_enabled (create_graph)
-                    before.copy_(after)
+                    # Add node meta for copy_ for partitioner that this node should be in backward graph.
+                    with torch.fx.traceback.preserve_node_meta(), set_partitioner_tag(
```
Nit on the name - I'd probably prefer something like "must be in backward".

Chillee reviewed (approved):
LGTM.

IvanKobzarev committed:
Update on "[aot] Keep backward mutations in backward"


https://github.com/pytorch/pytorch/issues/127561

Mutations of inputs in backward are emitted manually, after joint_fn tracing.
With default partitioner logic they will be moved to "forward" graph, as this is operation on forward inputs.

To keep those mutations in backward:
- Introduce "subgraph" node key, that can be specified with contextmanager. When we do manual `copy_` in backward on forward input - we know that his is for backward - set subgraph="backward"


In partitioner:
Introducing optional argument subgraph, to filter out nodes with specified subgraph (node_subgraph) and not to add them to subgraph if node_subgraph is different.


[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge
